### PR TITLE
No more slice removal for odd number of slices

### PR DESCRIPTION
**Pull request by MichielCottaar**
This is meant to address #217 . This pull request is still in draft, because the topup will crash for any images with odd number of slices. We still need to adjust the topup config file used (either for all input volumes or only for those with odd number of slices).

### CODE REVIEW

**Review by glasserm — COMMENTED** [review]
Thanks Michiel.  Is this tested?  It looks good from a code skim.

**Comment by MichielCottaar** [comment]
I'm running some overnight tests on HCP YA data. I'll let you know tomorrow

**Comment by MichielCottaar** [comment]
This is working, I've tested it in three specific cases. In all cases the input was the HCP YA adult data, which has an odd number of slices.
1. The default pipeline without any additional flags fails in that case, because the default topup configuration file is incompatible with an odd number of slices. I've copied the error message raised by topup below.
2. When adding the `--ensure-even-slices` flag, the pipeline removed a slice from both the input to topup and eddy. Topup ran succesfully afterward. I've not tested the rest of the pipeline, but see no reason why it would not.
3. When adding the `--topup-config-file=<file>` flag with a topup configuration without subsampling, topup also ran succesfully (without removing any slices).

The error message raised by topup if the subsampling levels are incompatible with the image dimensions
```
Part of FSL (ID: 6.0.6:e380285d)
topup

Usage: 
topup --imain=<some 4D image> --datain=<text file> --config=<text file with parameters> --out=my_topup_results


Compulsory arguments (You MUST set one or more of):
	--imain		name of 4D file with images
	--datain	name of text file with PE directions/times

Optional arguments (You may optionally specify one or more of):
	--out		base-name of output files (spline coefficients (Hz) and movement parameters)
	--fout		name of image file with field (Hz)
	--iout		name of 4D image file with unwarped images
	--logout	Name of log-file
	--warpres	(approximate) resolution (in mm) of warp basis for the different sub-sampling levels, default 10
	--subsamp	sub-sampling scheme, default 1
	--fwhm		FWHM (in mm) of gaussian smoothing kernel, default 8
	--config	Name of config file specifying command line arguments
	--miter		Max # of non-linear iterations, default 5
	--lambda	Weight of regularisation, default depending on --ssqlambda and --regmod switches. See user documetation.
	--ssqlambda	If set (=1), lambda is weighted by current ssq, default 1
	--regmod	Model for regularisation of warp-field [membrane_energy bending_energy], default bending_energy
	--estmov	Estimate movements if set, default 1 (true)
	--minmet	Minimisation method 0=Levenberg-Marquardt, 1=Scaled Conjugate Gradient, default 0 (LM)
	--splineorder	Order of spline, 2->Qadratic spline, 3->Cubic spline. Default=3
	--numprec	Precision for representing Hessian, double or float. Default double
	--interp	Image interpolation model, linear or spline. Default spline
	--scale		If set (=1), the images are individually scaled to a common mean, default 0 (false)
	--regrid		If set (=1), the calculations are done in a different grid, default 1 (true)
	-h,--help	display help info
	-v,--verbose	Print diagonostic information while running
	-h,--help	display help info



Topup: msg=topup_clp::topup_clp: Subsampling levels incompatible with image data.
Mon Jan 10 14:40:58 GMT 2022:run_topup.sh: While running '/home/fs0/ndcn0236/Work/datalad/HCP_test_slice_removal/Pipelines/DiffusionPreprocessing/scripts/run_topup.sh /home/fs0/ndcn0236/Work/datalad/HCP_test_slice_removal/base/103818/Diffusion/topup /home/fs0/ndcn0236/Work/datalad/HCP_test_slice_removal/Pipelines/global/config/b02b0.cnf':
Mon Jan 10 14:40:58 GMT 2022:run_topup.sh: ERROR: '/opt/fmrib/fsl/bin/topup' command failed with return code: 1
```

**Review by glasserm — CHANGES_REQUESTED** [review]
This is nearly there.  I do think the topup error message is actually not as userfriendly as we would like.  It would be better to catch the error early in the pipeline (before much processing has taken place) and to suggest the user to 1) Use a different config file (recommended) or 2) Use the --ensure-even-slices flag (if trying to replicate legacy behavior or if not using advanced eddy features).  Does that seem reasonable to you?

**Comment by MichielCottaar** [comment]
Sounds reasonable. I thought that I might need to parse the topup config file to do that. However, I just realised that we could simply raise an error if (1) there is an odd number of slices *and* (2) there is no `--topup-config-file` or `--ensure-even-slices` flag. I'll add an error in that case.

That would mean that if the user provides a topup config file that is incompatible with the data dimensions, they would still get the topup error. Is that okay with you?

**Comment by glasserm** [comment]
Is parsing the file hard?  @coalsont might be able to help with the code for this.  Basically it needs to grep the line: 
--subsamp=2,2,2,1,1 from the planned topup config and then check whether any digits are not 1.

**Comment by MichielCottaar** [comment]
It would take me quite a while to code that up in bash (I'm not very good in bash, grep, etc.). If someone can give me the code, I'll be happy to put it in.

**Comment by glasserm** [comment]
@coalsont is better than me, but I think it will be a one-liner.  Something like testing whether `cat <config> | grep "--subsamp=" | sed 's/--subsamp=//g' | sed 's/,//g' | sed 's/1//g'` returns a non-blank string.

**Comment by coalsont** [comment]
Honestly, if the user supplies their own topup config, I think they can deal with the errors they get from topup themselves, so I don't know that I would try to parse it to give an error in advance (maybe a future topup feature will allow subsampling without requirements on dimensions, which could make our error on someone else's config wrong and annoying).  I would recommend an option to ignore this check if we want our script to check custom config files by default.

If you want to try to protect users from their own custom config "mistakes" anyway, `sed 's/,//g'` is a bit brutal, and theoretically wouldn't catch "11", though hopefully nobody would use that.  Something like `| grep -Evq '^(1,)*1$'` on the end instead of the comma, 1 pair of `sed`s would give an exit status indication you could use in a conditional:

```bash
if grep -e "--subsamp=" <config> | sed 's/^.*--subsamp=//g' | grep -Evq '^[:space:]*(1,)*1[:space:]$' && [[ <slice count is odd> && "$EnsureEvenSlices" != "true" ]]
then
    log_Err_Abort "topup config uses subsampling, use --ensure-even-slices"
fi
```

If topup configs forbid whitespace, then some of the extra stuff in the regexps can be removed.

Technically, this assumes that the only kind of subsampling that will ever be in anyone's config is 2, since it assumes that an even number makes anything fine.  The current code also assumes that only the slice direction can be an issue.  Not sure how deep into this rabbit hole we really want to go, as long as we give the user the option of "just try it and see what topup thinks".  Really, the right solution is to ask FSL to improve topup's error messages.

**Comment by mharms** [comment]
Given that a 4x subsampling config file is now an option that FSL provides, and that the proposed check is only on the slice dimension, maybe it is indeed best to not get into the checking business at all (given that what we are proposing is not a complete check for the ways that the config file can be incorrect relative to the image dimensions).

But I defer to the consensus of others.

File: DiffusionPreprocessing/DiffPreprocPipeline.sh
Comment (mharms) [inline DiffusionPreprocessing/DiffPreprocPipeline.sh:138]:
```suggestion
                          File containing the FSL topup configuration.
```

File: DiffusionPreprocessing/DiffPreprocPipeline.sh
Comment (mharms) [inline DiffusionPreprocessing/DiffPreprocPipeline.sh:141]: Can we move the [--printcom] usage to the end of the usage statement?

**Comment by glasserm** [comment]
The topup error doesn't really tell the user how to resolve the problem as far as the HCP Pipelines are concerned.  If we wanted to catch and parse the topup error, that would be another option...

File: DiffusionPreprocessing/DiffPreprocPipeline.sh
Comment (mharms) [inline DiffusionPreprocessing/DiffPreprocPipeline.sh:156]:
```suggestion
                          because discarding a slice is incompatible with using slice-to-volume correction
```

File: DiffusionPreprocessing/DiffPreprocPipeline.sh
Comment (mharms) [inline DiffusionPreprocessing/DiffPreprocPipeline.sh:157]:
```suggestion
                          in FSL's eddy.
```

File: DiffusionPreprocessing/DiffPreprocPipeline_PreEddy.sh
Comment (mharms) [inline DiffusionPreprocessing/DiffPreprocPipeline_PreEddy.sh:96]:
```suggestion
                          File containing the FSL topup configuration.
```

File: DiffusionPreprocessing/DiffPreprocPipeline_PreEddy.sh
Comment (mharms) [inline DiffusionPreprocessing/DiffPreprocPipeline_PreEddy.sh:112]:
```suggestion
                          because discarding a slice is incompatible with using slice-to-volume correction
```

File: DiffusionPreprocessing/DiffPreprocPipeline_PreEddy.sh
Comment (mharms) [inline DiffusionPreprocessing/DiffPreprocPipeline_PreEddy.sh:113]:
```suggestion
                          in FSL's eddy.
```

**Comment by coalsont** [comment]
Detecting that topup crashed with a custom config and saying "if your config contains subsampling, make sure your image dimensions are an exact multiple" wouldn't require parsing anything.  When using the pipelines' own config files, we can provide more accurate errors in advance (because we know what they require), again without parsing anything.

File: DiffusionPreprocessing/DiffPreprocPipeline_PreEddy.sh
Comment (mharms) [inline DiffusionPreprocessing/DiffPreprocPipeline_PreEddy.sh:153]:
```suggestion
#  ${EnsureEvenSlices}    If "true" will ensure input images to topup/eddy have
```

File: DiffusionPreprocessing/DiffPreprocPipeline_PreEddy.sh
Comment (mharms) [inline DiffusionPreprocessing/DiffPreprocPipeline_PreEddy.sh:482]:
```suggestion
			log_Msg "Either supply a topup configuration file that doesn't use subsampling (e.g., FSL's 'b02b0_1.cnf') using the --topup-config-file=<file> flag (recommended)"
```

File: DiffusionPreprocessing/DiffPreprocPipeline_PreEddy.sh
Comment (mharms) [inline DiffusionPreprocessing/DiffPreprocPipeline_PreEddy.sh:484]:
```suggestion
			log_Msg "Note that the legacy option is incompatible with slice-to-volume correction in FSL's eddy"
```

**Review by mharms — COMMENTED** [review]
Looks good.  Thanks @MichielCottaar.  I've suggested some minor tweaks to some of the messages.

File: DiffusionPreprocessing/DiffPreprocPipeline.sh
Comment (MichielCottaar) [reply DiffusionPreprocessing/DiffPreprocPipeline.sh]: I've made this change for all diffusion preprocessing scripts.

**Comment by glasserm** [comment]
Ok

**Comment by mharms** [comment]
So, that does mean we are basically done, pending any further requests/suggestions from either @coalsont or @glasserm ?

**Comment by glasserm** [comment]
Well, I'd like things to be as userfriendly as possible, but don't want to hold things up if there isn't consensus on that.

**Comment by glasserm** [comment]
Shall we merge this so that we can tag a release once the tICA testing is done?

**Comment by mharms** [comment]
That's fine with me.

**Pull request merged**
